dll/win32/shdocvw/winlist.cpp

```diff
@@ -69,6 +82,367 @@ VariantClearLazy(_Inout_ LPVARIANTARG pvarg)
+    return S_OK;
+}
+
+STDMETHODIMP
+CConnectionPoint::GetConnectionPointContainer(
+    _Out_ IConnectionPointContainer **ppCPC)
+{
+    return m_pAnotherPoint->QueryInterface(IID_PPV_ARG(IConnectionPointContainer, ppCPC));
+}
+
+STDMETHODIMP
+CConnectionPoint::Advise(
+    _In_ IUnknown *pUnkSink,
+    _Out_ DWORD *pdwCookie)
+{
+    if (!pdwCookie)
+        return E_POINTER;
+
+    *pdwCookie = 0; // Zero is the invalid cookie
+
+    CComPtr<IDispatch> pDispatch;
+    HRESULT hr = pUnkSink->QueryInterface(*m_pIID, (PVOID *)&pDispatch);
+    if (FAILED(hr))
+    {
+        if (m_pIID != &IID_IPropertyNotifySink) // Windows mistook between pointer and body?
+            hr = pUnkSink->QueryInterface(IID_PPV_ARG(IDispatch, &pDispatch));
+        if (FAILED(hr))
+            return CONNECT_E_CANNOTCONNECT;
+    }
+
+    if (m_nItemNo >= m_nItems) // Needs grow?
+    {
+        INT cNewItems = m_nItems + 8;
+        DWORD cbNewItems = cNewItems * sizeof(CONNECTION);
+        PCONNECTION pNewConnections = (CONNECTION *)LocalReAllocHelp(m_pConnections, cbNewItems);
+        if (!pNewConnections)
+            return E_OUTOFMEMORY;
+        m_pConnections = pNewConnections;
+        m_nItems = cNewItems;
+    }
+
+    // Find an empty room
+    INT iConnect;
+    for (iConnect = 0; m_pConnections[iConnect].m_pDispatch; ++iConnect)
+    {
+        ;
+    }
+
+    m_pConnections[iConnect].m_pDispatch = pDispatch;
+    *pdwCookie = iConnect + 1; // This cookie is 1-based
+    ++m_nItemNo;
+
+    CComPtr<IConnectionPointCB> pCallback;
+    if (SUCCEEDED(m_pAnotherPoint->QueryInterface(IID_PPV_ARG(IConnectionPointCB, &pCallback))))
+        pCallback->Advise(*m_pIID, m_nItemNo, *pdwCookie);
+
+    return hr;
+}
+
+STDMETHODIMP
+CConnectionPoint::Unadvise(
+    _In_ DWORD dwCookie)
+{
+    if (!dwCookie)
+        return S_OK;
+
+    DWORD iItem = dwCookie - 1; // This cookie is 1-based
+    if (iItem >= m_nItems || !m_pConnections[iItem].m_pDispatch)
+        return CONNECT_E_NOCONNECTION;
+
+    CComPtr<IConnectionPointCB> pCallback;
+    if (SUCCEEDED(m_pAnotherPoint->QueryInterface(IID_PPV_ARG(IConnectionPointCB, &pCallback))))
+        pCallback->Unadvise(*m_pIID, m_nItemNo - 1, dwCookie);
+
+    if (m_pConnections[iItem].m_pDispatch)
+    {
+        m_pConnections[iItem].m_pDispatch->Release();
+        m_pConnections[iItem].m_pDispatch = NULL;
+    }
+
+    --m_nItemNo;
+    return S_OK;
+}
+
+STDMETHODIMP
+CConnectionPoint::EnumConnections(
+    _Out_ IEnumConnections **ppEnum)
+{
+    FIXME("\n");
+    return E_NOTIMPL;
+}
+
+STDMETHODIMP
+CConnectionPoint::DoInvokeIE4(
+    _In_ DWORD unknown1,
+    _In_ DWORD unknown2,
+    _In_ DISPID dispId,
+    _Inout_ DISPPARAMS *dispParams)
+{
+    return IConnectionPoint_InvokeWithCancel(this, dispId, dispParams, unknown1, unknown2);
+}
+
+STDMETHODIMP
+CConnectionPoint::SetRelativePath(
+    _In_ LPCWSTR pszPathRel,
+    _In_ DWORD dwReserved)
+{
+    return E_NOTIMPL; // Valid implementation
```

Comment:
why?

Reply:
At here, returning `E_NOTIMPL` is the correct implementation.

Reply:
The method names or parameter types might be wrong.

Reply:
Deleted.